rust/crd/src/authentication.rs

```diff
@@ -0,0 +1,290 @@
+    pub protocol: AuthenticationClassProtocol,
+}
+
+#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub enum AuthenticationClassProtocol {
+    Ldap(AuthenticationClassLdap),
+}
+
+#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct AuthenticationClassLdap {
+    /// Hostname of the LDAP server
+    pub hostname: String,
+    /// Port of the LDAP server
+    #[serde(default = "default_port")]
+    pub port: u16,
+    /// LDAP search base
+    #[serde(default)]
+    pub search_base: String,
+    /// LDAP query to filter users
+    #[serde(default)]
+    pub search_filter: String,
+    /// The name of the username field
+    #[serde(default = "default_uid_field")]
+    pub uid_field: String,
+    /// The name of the group field
+    #[serde(default = "default_group_field")]
+    pub group_field: String,
+    /// The name of the firstname field
+    #[serde(default = "default_firstname_field")]
+    pub firstname_field: String,
+    /// The name of the lastname field
+    #[serde(default = "default_lastname_field")]
+    pub lastname_field: String,
+    /// The name of the email field
+    #[serde(default = "default_email_field")]
+    pub email_field: String,
+    /// In case you need a special account for searching the LDAP server you can specify it here
+    pub bind_credentials: Option<AuthenticationClassLdapBindCredentials>,
+    /// Use a TLS connection. If not specified no TLS will be used
+    pub tls: Option<AuthenticationClassTls>,
+}
+
+fn default_port() -> u16 {
+    389
+}
+
+fn default_uid_field() -> String {
+    "uid".to_string()
+}
+
+fn default_group_field() -> String {
+    "memberof".to_string()
+}
+
+fn default_firstname_field() -> String {
+    "givenName".to_string()
+}
+
+fn default_lastname_field() -> String {
+    "sn".to_string()
+}
+
+fn default_email_field() -> String {
+    "mail".to_string()
+}
+
+#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct AuthenticationClassLdapBindCredentials {
+    /// [SecretClass](https://docs.stackable.tech/secret-operator/secretclass.html) containing the LDAP bind credentials
+    pub secret_class: String,
+    /// [Scope](https://docs.stackable.tech/secret-operator/scope.html) of the [SecretClass](https://docs.stackable.tech/secret-operator/secretclass.html)
+    pub scope: Option<AuthenticationClassSecretClassScope>,
+}
+
+#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct AuthenticationClassSecretClassScope {
+    #[serde(default)]
+    pub pod: bool,
+    #[serde(default)]
+    pub node: bool,
+    #[serde(default)]
+    pub services: Vec<String>,
+}
+
+#[derive(Clone, Debug, Deserialize, Eq, JsonSchema, PartialEq, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub enum AuthenticationClassTls {
```

Comment:
The PKI setup will be the same for all clients.. That said, on second thought I don't think `ConfigMap` or `Secret` make sense when `AuthenticationClass` isn't namespaced, since `Pod`s are only allowed to mount resources within their own NS.

I'd also drop `Path` since the meaning of that would be completely up to the individual product operators.

Reply:
Regarding Configmap/Secret
The usecase behind that was "I have an LDAP and a ca cert (only the cert, not the key)". I though the best way is to just throw the cert in a ConfigMap or Secret and reference it. I'm happy to remove it but struggeling to find another convenient way to simply provide the ca cert.

Regarding Path:
I also think this TLS struct will not only be used for LDAP but also for other stuff where it may make more sense. From 1&1 I'm all the Docker images (even when build on top of external images) install a deb/whatever package that puts the internal ca certs in a certain directory. For this usecase it would be needed to specify the cert path. I also think this is trivial to implement.

Reply:
CMs/Secrets: I can see that, but it feels like we need to think a bit harder about how to handle the mounting and RBAC issues. We'll have to come back to this eventually, I just think it makes sense to shelve this for the MVP until we have a more fully baked design.

Path: Is there a use-case that rebuilding all images with the cert baked-in would satisfy that CMs wouldn't? This feels like it's just an 1&1ism sneaking in that would confuse most users (and it's always easier to add it in later than to remove it...).

Reply:
If we remove CM and secret and path there is no way for the user to specify his ca cert. The only remaining option is the secret operator.
This renders the tls option useless for the most use-cases (where a corporate LDAP + ca exists) i guess. I think we need to support "just use this ca" in the first revision.

Reply:
They can still use a `k8sSearch` `SecretClass` that only matches the one secret with the CA certificate. The ergonomics aren't great (and we should improve them eventually), but it works today and it has defined cross-namespace semantics.

Reply:
We removed cm, secret and path as secret-operator can provide a ca cert without having the corresponding key and should enable all possible use-cases with well-defined behavior.
Question remaining is:
> I'm unsure whether it makes sense to have config in here, to me this could have been a simple enum insecure/server/twoway and the rest of the config like "ca this has been signed with" and "cas this accepts" as well as cipher suites etc. that we'll want to configure later I would have seen as external to this.

We moved the tls to tls.verification to leave space for additional attributes like cipher in the future